doc/programming/guidelines.rst

```diff
@@ -0,0 +1,291 @@
+        unused2: skip uint32;       # skip over 4 bytes
+
+        unused3: bytes &eod;        # extract and store remaining data
+        unused4: skip bytes &eod;   # skip over remaining data
+    };
+
+.. rubric::  Avoid Spicy strings and string manipulation
+
+Use Spicy :ref:`strings <type_string>` sparsely in your analyzer, and
+stick to :ref:`bytes <type_bytes>` instead where you can. Typically,
+you would convert ``bytes`` to strings as late as possible just before
+presentation to the user (including passing to Zeek events). Strings
+are never needed for the parsing process itself.
+
+As a corollary, avoid Spicy string manipulation. Always
+manipulate/concatenate ``bytes`` and convert only the final result to
+Spicy strings, probably in a ``%done`` hook or an EVT event. In
+particular, format strings come with a significant overhead. In other
+words, avoid use of ``%s`` to generate strings from ``bytes``.
+
+.. rubric::  Don't use temporary variables of expensive types just for readability
+
+Don't use temporary variables just to improve readability. In
+particular, ``string`` and ``bytes`` (that are implemented as C++
+strings under the hood and can be expensive to use) need to be
+allocated and destroyed. This may introduce relevant overhead as it
+cannot be guaranteed that the C++ compiler will be able to optimize
+the code generated by Spicy. To improve readability, comments are the
+tool of choice. In theory, benchmarking the impact of using a
+temporary variable is a valid approach as well. However, the benefit
+of slightly increased readability most likely does not justify the
+effort of benchmarking.
+
+.. rubric::  Remove unnecessary hooks
+
+Multiple hook handlers with different priorities can be defined in
+various places like inside a unit or, with Zeek, in EVT files.
+However, hooks (either unit or field) should be avoided when not
+needed. Using hooks comes with a performance cost because of
+additional code generated by the compiler, which executes during
+parsing. While ``%init`` hooks are often used for initializing unit
+variables, we can often eliminate them by providing default values in
+the variable definition:
+
+.. code::
+
+    public type Message = unit {
+
+        on %init {
+            self.A1 = 23;     # explicit initialization through hook
+        }
+
+        var A1: uint32;
+        var A2: uint32 = 23;  # implicit initialization through default value
+    };
+
+.. rubric::  Avoid recursion
+
+The Spicy compiler allows declaration of recursive units with runtime
+conditions dictating when the recursion terminates. However,
+benchmarking has revealed that recursion introduces much higher
+overheads as compared to unrolled linear/iterative code performing
+similar functionality. This overhead is due to increased lifetimes of
+the units, data and their associated hooks; as well as less potential
+for compiler optimization due to additional internal machinery around
+the recursive calls.
```

Comment:
I would rephrase it to remove the "revelations" (which would seem weird for an in-tree doc),

```suggestion
The Spicy allows declaration of recursive units with runtime conditions
dictating when the recursion terminates. A recursive grammar dictates that
units higher up in the recursion chain to be kept around until the overall
recursion terminates. This incurs additional memory overhead and might reduce
potential for optimizations. A unrolled linear/iterative implementation does
not have these issues.
```

Reply:
I'll try to tweak it a bit differently.